### PR TITLE
perf: Utilize ORDER BY LIMIT over ROW_NUMBER where possible

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/rewrite.py [hunk 1]
```diff
@@ -385,6 +385,44 @@ def common_selection_root(
     return None
 
 
+def pullup_limit_from_slice(
```
Comment (chelsea-lin) [inline bigframes/core/rewrite.py:388]: Are we able to have a test to verify this process?

File: bigframes/core/rewrite.py [hunk 2]
```diff
@@ -385,46 +385,72 @@ def common_selection_root(
...
+        return rewrite_slice(cast(nodes.SliceNode, root))
     else:
         return root.transform_children(replace_slice_ops)
 
 
-def get_simplified_slice(node: nodes.SliceNode):
-    """Attempts to simplify the slice."""
-    row_count = traversals.row_count(node)
-    start, stop, step = node.start, node.stop, node.step
+def rewrite_slice(node: nodes.SliceNode):
```
Comment (tswast) [inline bigframes/core/rewrite.py:429]: `rewrite_slice` seems like a good candidate for some unit tests if we don't have some already.

File: bigframes/session/executor.py [hunk 3]
```diff
@@ -575,20 +572,8 @@ def _validate_result_schema(
 
 
 def generate_head_plan(node: nodes.BigFrameNode, n: int):
-    offsets_id = bigframes.core.guid.generate_guid("offsets_")
-    plan_w_offsets = nodes.PromoteOffsetsNode(
-        node, bigframes.core.identifiers.ColumnId(offsets_id)
-    )
-    predicate = ops.lt_op.as_expr(ex.deref(offsets_id), ex.const(n))
-    plan_w_head = nodes.FilterNode(plan_w_offsets, predicate)
     # Finally, drop the offsets column
```
Comment (tswast) [inline bigframes/session/executor.py:575]: I think this comment is out of date now.

File: bigframes/core/compile/compiled.py [hunk 4]
```diff
@@ -969,7 +970,10 @@ def to_sql(
                 order_by_clause = bigframes.core.sql.ordering_clause(
                     baked_ir._ordering.all_ordering_columns
                 )
-                sql += f"{order_by_clause}\n"
+                sql += f"\n{order_by_clause}"
+            if limit is not None:
+                assert isinstance(limit, int)
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py:975]: Yeah, in general, probably best to assume that wrong types can sometimes reach even the lower levels of the code base. In particular want to be safe about what we are putting into the sql strings. Switched to TypeError here.

File: bigframes/core/nodes.py [hunk 5]
```diff
@@ -304,6 +310,20 @@ def variables_introduced(self) -> int:
     def relation_ops_created(self) -> int:
         return 2
 
+    @property
+    def is_limit(self) -> bool:
+        """Returns whether this is equivalent to a ORDER BY ... LIMIT N."""
+        return (not self.start) and (self.step == 1) and (self.stop is not None)
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/nodes.py:316]: yeah, should be doable, if a bit tricky. adding a todo

File: bigframes/core/nodes.py [hunk 6]
```diff
@@ -351,6 +371,15 @@ def variables_introduced(self) -> int:
...
         return True
 
+    @property
+    def row_count(self) -> Optional[int]:
+        if self.type == "cross":
+            if self.left_child.row_count is None or self.right_child.row_count is None:
+                return None
+            return self.left_child.row_count * self.right_child.row_count
+
+        return None
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/nodes.py:381]: Need a bit more than just unique join keys to determine exact cardinality. The join keys on the outer side all need to have a match on the inner side as well. Most of these cases will be handled by implicit joiner.

I think upper bound will also be useful (and possible to generate for every node type except explode), so may add that as well later on.

File: bigframes/core/nodes.py [hunk 7]
```diff
@@ -635,12 +681,24 @@ def relation_ops_created(self) -> int:
         return 3
 
     @property
-    def supports_fast_head(self) -> bool:
+    def fast_offsets(self) -> bool:
         # Fast head is only supported when row offsets are available.
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/nodes.py:685]: Updated comment to mention clustering

File: bigframes/core/nodes.py [hunk 8]
```diff
@@ -635,12 +681,24 @@ def relation_ops_created(self) -> int:
...
+            return False
+        order_cols = self.source.ordering.all_ordering_columns
+        # monotonicity would probably be fine
+        if not all(col.scalar_expression.is_identity for col in order_cols):
+            return False
+        order_col_ids = tuple(
+            cast(ex.DerefOp, col.scalar_expression).id.name for col in order_cols
+        )
+        cluster_col_ids = self.source.table.cluster_cols
+        return order_col_ids == cluster_col_ids
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/nodes.py:701]: I think its best to be conservative on this. In theory, the important parts of the order ids could be the later parts, not in the clustering. Also, the query engine is binary about this optimization, either on or off. Can maybe run some tests and loosen the requirement later.

File: bigframes/session/executor.py (hunk 3 above)
Comment (TrevorBergeron) [reply bigframes/session/executor.py]: removed

File: bigframes/core/rewrite.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/core/rewrite.py:428]: ok, added some unit tests. Bit rough writing tree tests right now. Might need to setup a little framework if we want to do lots of these. Local engine could also be a way to ensure that tree rewrites do not change result value.

**Review by tswast — APPROVED** [review]
Love it!

Do we have data on how this will affect our benchmark results?

**Comment by TrevorBergeron** [comment]
> Love it!
> 
> Do we have data on how this will affect our benchmark results?

Probably most impact on notebook benchmarks initially. TPCH uses to_gbq which will not use this optimization until further changes (need to be careful not to use ORDER BY LIMIT when writing to clustered tables)

**Pull request merged**